### ISSUE TITLE
Reproduce the word analogy result with WikiPedia dataset from "Enriching word ..." paper

### ISSUE THREAD

**Issue by junjieqian** [facebookresearch/fastText#563]
Hi, I am reading the paper "Enriching word vectors with subword information" but cannot repro the word analogy result with fastText for English Wikipedia dataset (and Mikolov's training dataset). The result I got is far lower than the one reported in the paper.

After looking at the result, I found that there are a lot of "weird" words in the vocabulary which leads to wrong prediction results. For example, the vocabulary contains "works—and", "works,", "works»", "♥♥talk" and some other non-english words. These special words show in the normalized wikipedia dataset (which is feed as training dataset later) as well.

This script is used to download and parse the Wikipedia dataset, https://github.com/facebookresearch/fastText/blob/master/get-wikimedia.sh, and as the paper instructs it normalize the dataset as well. Also, seems this line of code, https://github.com/facebookresearch/fastText/blob/25d0bb04bf43d8b674fe9ae5722ef65a0856f5d6/src/dictionary.cc#L160, computes all such non-ascii chars in the word and adds to vocabulary. Is this by design for English words?

Is there any other steps I missed? Or we need to manually ignore the special chars?
Regarding the training, the parameters are set accordingly to the paper, `-lr 0.025 -dim 300 -ws 5 -epoch 10 -minCount 5 -neg 5 -loss ns -bucket 2000000 -minn 3 -maxn 6 -thread 16 -t 1e-4 -lrUpdateRate 100`

Thank you for your time!

**Comment by EdouardGrave** [comment]
Hi @junjieqian,

Thank you for reporting this issue.

What kind of results do you obtain? Are you evaluating these vectors with the `compute-accuracy` tool from word2vec?

Best,
Edouard.

**Comment by danieldk** [comment]
*, computes all such non-ascii chars in the word and adds to vocabulary. Is this by design for English words?*

Note that `word[j] & 0xc0 == 0x80` only holds for UTF-8 continuation bytes. Without this loop, some n-grams would be invalid UTF-8 sequences (ending in a single byte of a character with a multi-byte encoding).

**Comment by junjieqian** [comment]
Hi @EdouardGrave and @danieldk  , thanks for the reply!
For the work analogy result, I only got 509 correct ones. Some of the incorrect ones are due to some non-ascii codes in the predicted ones (e.g. "worker ," instead of "worker"). I did not use `compute-accuracy` in word2vec for the evaluation, but directly fed the words sequence to `fastText`.

I agree that `word[j] & 0xc0 == 0x80` is needed to avoid invalid UTF-8 sequences but not sure if there is a better way to get rid of non-ascii chars ("," or so on).

Thanks

**Comment by EdouardGrave** [comment]
Hi @junjieqian,

How do you use fastText to compute the word analogy results? As of now, fastText cannot be used directly to compute the results on the Mikolov et al. dataset.

Have you tried running the `compute-accuracy` tool from word2vec?
```
./compute-accuracy MODEL.VEC 200000 < question-words.txt
```
The result from this tool would be helpful to resolve this issue.

In order to get a "cleaner" training set, you can use the tokenizer from [Moses](https://github.com/moses-smt/mosesdecoder). To do so, you can replace `normalize_text` by `perl mosesdecoder/scripts/tokenizer/tokenizer.perl -no-escape` in the `get-wikimedia.sh` script.

Best,
Edouard

**Comment by junjieqian** [comment]
Hi @EdouardGrave , thank you very much for your time and attention on this issue.

I did some change in fastText to make it read in a file (each line contains `A/B/C`) and write the output (`D`) to a file.

The output of `compute-accuracy` tool from word2vec is as following. It does not look correct. But even with the pretrained models (from fastText website), the results are same.

```
capital-common-countries:
ACCURACY TOP1: 0.00 %  (0 / 1)
Total accuracy: -nan %   Semantic accuracy: -nan %   Syntactic accuracy: -nan %
capital-world:
ACCURACY TOP1: 0.00 %  (0 / 1)
Total accuracy: -nan %   Semantic accuracy: -nan %   Syntactic accuracy: -nan %
currency:
ACCURACY TOP1: 0.00 %  (0 / 1)
Total accuracy: -nan %   Semantic accuracy: -nan %   Syntactic accuracy: -nan %
city-in-state:
ACCURACY TOP1: 0.00 %  (0 / 1)
Total accuracy: -nan %   Semantic accuracy: -nan %   Syntactic accuracy: -nan %
family:
ACCURACY TOP1: 0.00 %  (0 / 2)
Total accuracy: 0.00 %   Semantic accuracy: 0.00 %   Syntactic accuracy: -nan %
gram1-adjective-to-adverb:
ACCURACY TOP1: 0.00 %  (0 / 1)
Total accuracy: 0.00 %   Semantic accuracy: 0.00 %   Syntactic accuracy: -nan %
gram2-opposite:
ACCURACY TOP1: 0.00 %  (0 / 1)
Total accuracy: 0.00 %   Semantic accuracy: 0.00 %   Syntactic accuracy: -nan %
gram3-comparative:
ACCURACY TOP1: 0.00 %  (0 / 1)
Total accuracy: 0.00 %   Semantic accuracy: 0.00 %   Syntactic accuracy: -nan %
gram4-superlative:
ACCURACY TOP1: 0.00 %  (0 / 1)
Total accuracy: 0.00 %   Semantic accuracy: 0.00 %   Syntactic accuracy: -nan %
gram5-present-participle:
ACCURACY TOP1: 0.00 %  (0 / 1)
Total accuracy: 0.00 %   Semantic accuracy: 0.00 %   Syntactic accuracy: -nan %
gram6-nationality-adjective:
ACCURACY TOP1: 0.00 %  (0 / 1)
Total accuracy: 0.00 %   Semantic accuracy: 0.00 %   Syntactic accuracy: -nan %
gram7-past-tense:
ACCURACY TOP1: 0.00 %  (0 / 1)
Total accuracy: 0.00 %   Semantic accuracy: 0.00 %   Syntactic accuracy: -nan %
gram8-plural:
ACCURACY TOP1: 0.00 %  (0 / 2)
Total accuracy: 0.00 %   Semantic accuracy: 0.00 %   Syntactic accuracy: 0.00 %
gram9-plural-verbs:
ACCURACY TOP1: 0.00 %  (0 / 1)
Total accuracy: 0.00 %   Semantic accuracy: 0.00 %   Syntactic accuracy: 0.00 %
Questions seen / total: 4 19544   0.02 %
```

**Comment by EdouardGrave** [comment]
Hi @junjieqian,

Please note that A and B should be inverted between the Mikolov et al. dataset and the fastText implementation of `analogies`.

I forgot to mention, but you need to make two small changes in `compute-accuracy.c` so that it can read text files instead of binary ones. First, line 64 should be:
```
for (a = 0; a < size; a++) fscanf(f, "%f", &M[a + b * size]);
```
instead of
```
for (a = 0; a < size; a++) fread(&M[a + b * size], sizeof(float), 1, f);
```
Second, you need to add:
```
char c = fgetc(f);
``` 
after line 68 (before the end of the `for` loop).

Best,
Edouard
